apps/heft/src/plugins/CopyFilesPlugin.ts

```diff
@@ -71,86 +66,43 @@ interface ICopyDescriptor {
+  terminal: ITerminal,
+  watchFileSystemAdapter?: WatchFileSystemAdapter
 ): Promise<void> {
-  const copyDescriptors: ICopyDescriptor[] = await _getCopyDescriptorsAsync(
+  const copyDescriptors: Map<string, ICopyDescriptor> = await _getCopyDescriptorsAsync(
     copyOperations,
-    // Use the normal globber if it is the first run, to ensure that non-watched files are copied
-    isFirstRun ? glob : globChangedFilesAsyncFn
+    watchFileSystemAdapter
   );
-  await _copyFilesInnerAsync(copyDescriptors, logger);
+
+  await _copyFilesInnerAsync(copyDescriptors, terminal);
 }
 
 async function _getCopyDescriptorsAsync(
-  copyConfigurations: ICopyOperation[],
-  globFn: GlobFn
-): Promise<ICopyDescriptor[]> {
-  const processedCopyDescriptors: ICopyDescriptor[] = [];
-
+  copyConfigurations: Iterable<ICopyOperation>,
+  fs: WatchFileSystemAdapter | undefined
+): Promise<Map<string, ICopyDescriptor>> {
   // Create a map to deduplicate and prevent double-writes
   // resolvedDestinationFilePath -> descriptor
   const destinationCopyDescriptors: Map<string, ICopyDescriptor> = new Map();
 
   await Async.forEachAsync(
     copyConfigurations,
     async (copyConfiguration: ICopyOperation) => {
-      let sourceFolder: string | undefined;
-      let sourceFilePaths: Set<string> | undefined;
-      if (
-        !copyConfiguration.fileExtensions?.length &&
-        !copyConfiguration.includeGlobs?.length &&
-        !copyConfiguration.excludeGlobs?.length
-      ) {
-        sourceFolder = path.dirname(copyConfiguration.sourcePath);
-        if (copyConfiguration.hardlink) {
-          // Specify a glob to match all files in the folder, since folders cannot be hardlinked.
-          // Perform globbing from one folder up, so that we create the folder in the destination.
-          try {
-            sourceFilePaths = await getFilePathsAsync(
-              {
-                ...copyConfiguration,
-                sourcePath: sourceFolder,
-                includeGlobs: [`${path.basename(copyConfiguration.sourcePath)}/**/*`]
-              },
-              globFn
-            );
-          } catch (error) {
-            if (FileSystem.isNotDirectoryError(error)) {
-              // The source path is a file, not a folder. Handled below.
-            } else {
-              throw error;
-            }
-          }
-        }
-
-        // Still not set, either it's not a hardlink or it's a file.
-        if (!sourceFilePaths) {
-          sourceFilePaths = new Set([copyConfiguration.sourcePath]);
-        }
-      } else {
-        // Assume the source path is a folder
-        sourceFolder = copyConfiguration.sourcePath;
-        sourceFilePaths = await getFilePathsAsync(copyConfiguration, globFn);
-      }
+      // The source path is required to be a folder
+      const sourceFolder: string | undefined = copyConfiguration.sourcePath;
+      const sourceFilePaths: Set<string> | undefined = await getFilePathsAsync(copyConfiguration, fs);
```

Comment:
Why remove the parts allowing for files/folders without globs to be specified in the sourcePath? I thought that was a nice addition.